pkg/policy/manager.go

```diff
@@ -0,0 +1,655 @@
+		esClient:      esClient,
+		esConfig:      esConfig,
+		indexManager:  indexManager,
+		filterer:      filterer,
+		opa:           opa,
+		grafeasCommon: grafeasCommon,
+	}
+}
+
+var newUuid = uuid.New
+
+func (m *manager) CreatePolicy(ctx context.Context, policy *pb.Policy) (*pb.Policy, error) {
+	log := m.logger.Named("CreatePolicy")
+	log.Debug("received request", zap.Any("request", policy))
+
+	if len(policy.Name) == 0 {
+		return nil, status.Errorf(codes.InvalidArgument, "policy name not provided")
+	}
+
+	policyId := newUuid().String()
+	log = log.With(zap.String("id", policyId))
+
+	version := int32(1)
+	policyVersionId := policyVersionId(policyId, version)
+	currentTime := timestamppb.Now()
+
+	policy.Id = policyId
+	policy.Created = currentTime
+	policy.Updated = currentTime
+	policy.CurrentVersion = version
+
+	policyVersion := policy.Policy
+	policy.Policy = nil // remove current policy
+	policyVersion.Version = version
+	policyVersion.Message = "Initial policy creation"
+	policyVersion.Created = currentTime
+
+	log.Debug("validating policy")
+	result, err := m.ValidatePolicy(ctx, &pb.ValidatePolicyRequest{Policy: policyVersion.RegoContent})
+	if (err != nil) || !result.Compile {
+		message := &pb.ValidatePolicyResponse{
+			Policy:  policyVersion.RegoContent,
+			Compile: false,
+			Errors:  result.Errors,
+		}
+		s, _ := status.New(codes.InvalidArgument, "failed to compile the provided policy").WithDetails(message)
+		return nil, s.Err()
+	}
+
+	log.Debug("performing bulk request")
+	response, err := m.esClient.Bulk(ctx, &esutil.BulkRequest{
+		Index:   m.policiesAlias(),
+		Refresh: m.esConfig.Refresh.String(),
+		Items: []*esutil.BulkRequestItem{
+			{
+				Operation:  esutil.BULK_CREATE,
+				Message:    policy,
+				DocumentId: policyId,
+				Join: &esutil.EsJoin{
+					Field: policyDocumentJoinField,
+					Name:  policyRelationName,
+				},
+			},
+			{
+				Operation:  esutil.BULK_CREATE,
+				Message:    policyVersion,
+				DocumentId: policyVersionId,
+				Join: &esutil.EsJoin{
+					Parent: policyId,
+					Field:  policyDocumentJoinField,
+					Name:   policyVersionRelationName,
+				},
+			},
+		},
+	})
+
+	if err != nil {
+		return nil, createError(log, "error creating policy", err)
+	}
+
+	var bulkErrors []error
+	for _, item := range response.Items {
+		if item.Create.Error != nil {
+			bulkErrors = append(bulkErrors, fmt.Errorf("error creating policy [%d] %s: %s", item.Create.Status, item.Create.Error.Type, item.Create.Error.Reason))
+		}
+	}
+
+	if len(bulkErrors) > 0 {
+		return nil, createError(log, "failed to create policy", fmt.Errorf("bulk creation errors: %v", bulkErrors))
+	}
+
+	policy.Policy = policyVersion
+
+	log.Debug("successfully created policy")
+	return policy, nil
+}
+
+func (m *manager) GetPolicy(ctx context.Context, request *pb.GetPolicyRequest) (*pb.Policy, error) {
+	log := m.logger.Named("GetPolicy")
+	log.Debug("received request")
+	policyId, version, err := parsePolicyVersionId(request.Id)
```

Comment:
This works but I think there might be a simpler solution. Let's pair a little tomorrow and see what we can come up with.

Reply:
Definitely open to another approach. Posting for a few others here for tomorrow:

1. Check the length and then index into it:
```go
id := request.Id[0:35]
version := request.Id[37:]
```
2. Use a regex:
```go
r := regexp.MustCompile(`^(?P<id>[0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12}).?(?P<version>\d?)$`)

matches := r.FindStringSubmatch(request.Id)

id := matches[r.SubexpIndex("id")]
version := matches[r.SubexpIndex("version")]
```
3. Send it off to Elasticsearch and check the document type or join field